### DISCUSSION TITLE
`GeoShapePreIndexed` missing optional parameter `routing`

### DISCUSSION THREAD

**Discussion by maxxer** [ruflin/Elastica#2217] [Q&A]
I just filed issue https://github.com/ruflin/Elastica/issues/2216, because I believe it misses the ability to declare the optional parameter `routing`. To my understanding, it's not possible to add it with the currently available methods, because the `toArray()` produces a _static_ structure that doesn't include that attribute.

Is there any other way than altering the class, that I didn't spot? thanks

**Comment by sidz** [comment]
Hey @maxxer 

You always could write and use your own `GeoShapePreIndexed` class which could extend `GeoShapePreIndexed` or `AbstractGeoShape`

For example:
```php
class GeoShapePreIndexed extends AbstractGeoShape
{
    private $_indexedId;

    private $_indexedIndex;

    private $_indexedPath;

    private $_routing;

    public function __construct(
        string $path,
        string $indexedId,
        string $indexedIndex,
        string $indexedPath,
        ?string $routing = null,
    ) {
        $this->_path = $path;
        $this->_indexedId = $indexedId;
        $this->_indexedIndex = $indexedIndex;
        $this->_indexedPath = $indexedPath;
        $this->_routing = $routing;
    }

    /**
     * {@inheritdoc}
     */
    public function toArray(): array
    {
        return [
            'geo_shape' => [
                $this->_path => [
                    'indexed_shape' => array_filter([
                        'id' => $this->_indexedId,
                        'index' => $this->_indexedIndex,
                        'path' => $this->_indexedPath,
                        'routing' => $this->_routing,
                    ]),
                    'relation' => $this->_relation,
                ],
            ],
        ];
    }
}
```

**Reply by maxxer** [reply]
I was looking for other class methods to add the parameter, but indeed this is a solution

**Reply by ruflin** [reply]
@maxxer Saw this conversation too late. Can you follow up on the Github issue you created?
